File: workspace/judoku/src/main/java/judoku/data/Cell.java
package judoku.data;

import java.util.HashSet;
import java.util.Observable;
import java.util.Set;

import judoku.Debug;

public class Cell extends Observable {
	private OneToNine value = null;
	private Set<OneToNine> potentialValues = null;
	public final int id;

	// reference to the cell's line
	private Line line;
	// reference to the cell's block
	private Block block;
	// reference to the cell's column
	private Column column;

	public Cell(Grid grid, int id) {
		this.id=id;

		// initialize the cell with a list of potential values from 1..9
		potentialValues = new HashSet<OneToNine>();
		for (OneToNine o2n: OneToNine.values()) {
			potentialValues.add(o2n);
		}
	}

	// initialize the cells
	void parseCharCell(char c) {
		setValue(OneToNine.fromChar(c));
	}

	// -----------------------------------------------------------------------------------------------

	public Block getBlock() {
		return block;
	}
	public void setBlock(Block block) {
		this.block = block;
	}

	public Column getColumn() {
		return column;
	}
	public void setColumn(Column column) {
		this.column = column;
	}

	public Line getLine() {
		return line;
	}
	public void setLine(Line line) {
		this.line = line;
	}

	// -----------------------------------------------------------------------------------------------

	// get the cells value (if fixed)
	public OneToNine getValue() {
		return value;
	}

	// get the list of potential values (never null, but might be empty)
	public Set<OneToNine> getPotentialValues() {
		return potentialValues;
	}

	// set the list of potential values (never null, but might be empty)
	public void setPotentialValues(Set<OneToNine> potentialValues) {
		this.potentialValues = potentialValues;
	}

	// fix value and then notify all observers (ie line, block, column of the cell)
	public void setValue(OneToNine value) {
		if (value == null) return;
		
		if (Debug.FIXED_DEBUG) {
			System.out.println("Fix cell " + id + " to value "+ value);
		}
		this.value = value;
		potentialValues.clear();	// better than null to prevent NPEs elsewhere
		setChanged();
		notifyObservers();
	}

	// delete a value from the list of potential values
	public void deleteValue(OneToNine deleteValue) {
		if (isFix()) return;
		if (Debug.DELETE_DEBUG) {
			System.out.println("Delete value "+ deleteValue + " from cell " + id);
		}
		this.getPotentialValues().remove(deleteValue);
	}

	// -----------------------------------------------------------------------------------------------

	// is this cell fixed already?
	public boolean isFix() {
		return value != null;
	}

	// -----------------------------------------------------------------------------------------------
	
	public String toString() {
		StringBuffer buf = new StringBuffer();

		if (!Debug.CELL_DEBUG) {
			if (isFix()) {
				buf.append(value.toString());
			}
			else {
				buf.append(".");
			}
		}
		else {
			if (id<10) buf.append(" ");
			buf.append(" " + id);
			if (isFix()) {
				buf.append("[");
				buf.append("........");
				buf.append(value.toString());
				buf.append("]");
			}
			else {
				buf.append("{");
				int numBlanks = 9 - (potentialValues.size());
				for (int i=0; i<numBlanks; i++) buf.append(".");
				for (OneToNine o2n: potentialValues) {
					buf.append(o2n.toString());
				}
				buf.append("}");
			}
		}

		return buf.toString();
	}

	public boolean equals(Object o) {
		if (o==null) return false;
		if (o instanceof Cell) {
			Cell oAsCell = (Cell) o;
			return (this.id == oAsCell.id);
		}
		return false;
	}

	public int hashCode () {
		return id;
	}
}

File: workspace/scudoku/src/main/scala/scudoku/data/Cell.scala
package scudoku.data

/**
 * Java2Scala port from judoku.data.Cell
 */

import UnknownOrOneToNine._
import java.util.Observable
import scudoku.Debug

// TODO implement Observable as in https://stackoverflow.com/questions/13435151/implementing-observer-pattern
class Cell (val grid: Grid, val block: Block, val line: Line, val column: Column, val id: Int) extends Observable { 
  //TODO can we implement as a case class? (we get equals/hashCode/toString for free)
    
  private var _value: UnknownOrOneToNine = UNKNOWN
  
  // get the cell's value (if fixed)
  def value = _value 

  // TODO override did not work - had to rename cellValue to _cellValue 
  // fix value and then notify all observers (ie line, block, column of the cell)
  def value_=(newValue: UnknownOrOneToNine): Unit = {  
    _value = newValue
    for (potentialValue <- potentialValues) potentialValues-=(potentialValue)
    
    if (Debug.FIXED_DEBUG) {
      Debug.debug("Fix cell " + id + " to value " + value);
    }
    
    // Observable
    setChanged
    notifyObservers
  }
  
  // list of potential values (never null, but might be empty)
  val potentialValues = scala.collection.mutable.Set[UnknownOrOneToNine](ONE, TWO, THREE, FOUR, FIVE, SIX, SEVEN, EIGHT, NINE)
    
  // delete a value from the list of potential values
  def deletePotentialValue(valueToDelete: UnknownOrOneToNine) {
    potentialValues-=(valueToDelete)
    
    if (Debug.DELETE_DEBUG) {
       Debug.debug("Delete value "+ valueToDelete + " from cell " + id); 
    }
  }
  
  // is this cell fixed already?
  def isFix: Boolean = _value != UNKNOWN 
  
  // checks if cell and otherCell have the same fix value (and are different objects)
  def isDuplicate(otherCell: Cell): Boolean = {
    this != otherCell && 
      isFix && otherCell.isFix && 
      value==otherCell.value
  }
    
  // --------------------------------------------------------

  override def equals(o: Any): Boolean = {
    o match {
      case o: Cell => this.id == o.id
      case _ => false
    }
  }
  override def hashCode: Int = id
  
  // --------------------------------------------------------
  
  override def toString: String = {
    val buf: java.lang.StringBuffer = new StringBuffer()
    
    if (! Debug.CELL_DEBUG) {
      if (isFix) {
        buf.append(value.toString());
      }
      else {
        buf.append(".");
      }
     }
     else {
       if (id<10) buf.append(" ")
       buf.append(id)
       if (isFix) {
          buf.append("[");
          buf.append("________");
          buf.append(_value);
          buf.append("] ");
        }
        else {
          buf.append("{");
          val numBlanks = 9 - potentialValues.size
          for (i <- 0 to numBlanks-1) buf.append(".");
          for (o2n <- potentialValues) {
            buf.append(o2n);
          }
          buf.append("} ");
        }    
    }
    buf.toString()
  }
}